app/app.go

```diff
@@ -371,17 +384,22 @@ func NewOsmosisApp(
 		gamm.NewAppModule(appCodec, app.GAMMKeeper),
 		incentives.NewAppModule(appCodec, app.IncentivesKeeper),
 		lockup.NewAppModule(appCodec, app.LockupKeeper),
+		epochs.NewAppModule(appCodec, app.EpochsKeeper),
 	)
 
 	// During begin block slashing happens after distr.BeginBlocker so that
 	// there is nothing left over in the validator fee pool, so as to keep the
 	// CanWithdrawInvariant invariant.
 	// NOTE: staking module is required if HistoricalEntries param > 0
 	app.mm.SetOrderBeginBlockers(
+		epochstypes.ModuleName,
 		upgradetypes.ModuleName, minttypes.ModuleName, distrtypes.ModuleName, slashingtypes.ModuleName,
 		evidencetypes.ModuleName, stakingtypes.ModuleName, ibchost.ModuleName,
 	)
-	app.mm.SetOrderEndBlockers(crisistypes.ModuleName, govtypes.ModuleName, stakingtypes.ModuleName)
+	app.mm.SetOrderEndBlockers(
+		crisistypes.ModuleName, govtypes.ModuleName, stakingtypes.ModuleName,
+		epochstypes.ModuleName,
```

Comment:
Shouldn't epochs module come first if it handles the epochs for everyone else?

Reply:
EndBlocker = end of epochs, and epochs endblock should be "real" end of epochs - after all modules execution.

Reply:
I see, can a comment be added here to explain this?

Reply:
Added comment